module/zcommon/zfs_fletcher_aarch64_neon.c

```diff
@@ -60,141 +60,185 @@ static void
 fletcher_4_aarch64_neon_fini(fletcher_4_ctx_t *ctx, zio_cksum_t *zcp)
 {
 	uint64_t A, B, C, D;
-	A = ctx->aarch64_neon[0].v[0] + ctx->aarch64_neon[0].v[1];
-	B = 2 * ctx->aarch64_neon[1].v[0] + 2 * ctx->aarch64_neon[1].v[1] -
-	    ctx->aarch64_neon[0].v[1];
-	C = 4 * ctx->aarch64_neon[2].v[0] - ctx->aarch64_neon[1].v[0] +
-	    4 * ctx->aarch64_neon[2].v[1] - 3 * ctx->aarch64_neon[1].v[1];
-	D = 8 * ctx->aarch64_neon[3].v[0] - 4 * ctx->aarch64_neon[2].v[0] +
-	    8 * ctx->aarch64_neon[3].v[1] - 8 * ctx->aarch64_neon[2].v[1] +
-	    ctx->aarch64_neon[1].v[1];
+
+	A = ctx->aarch64_neon[0].v[0] + ctx->aarch64_neon[0].v[1] +
+	    ctx->aarch64_neon[0].v[2] + ctx->aarch64_neon[0].v[3];
+	B = 0 - ctx->aarch64_neon[0].v[1] - 2 * ctx->aarch64_neon[0].v[2] -
+	    3 * ctx->aarch64_neon[0].v[3] + 4 * ctx->aarch64_neon[1].v[0] +
+	    4 * ctx->aarch64_neon[1].v[1] + 4 * ctx->aarch64_neon[1].v[2] +
+	    4 * ctx->aarch64_neon[1].v[3];
+
+	C = ctx->aarch64_neon[0].v[2] + 3 * ctx->aarch64_neon[0].v[3] -
+	    6 * ctx->aarch64_neon[1].v[0] - 10 * ctx->aarch64_neon[1].v[1] -
+	    14 * ctx->aarch64_neon[1].v[2] - 18 * ctx->aarch64_neon[1].v[3] +
+	    16 * ctx->aarch64_neon[2].v[0] + 16 * ctx->aarch64_neon[2].v[1] +
+	    16 * ctx->aarch64_neon[2].v[2] + 16 * ctx->aarch64_neon[2].v[3];
+
+	D = 0 - ctx->aarch64_neon[0].v[3] + 4 * ctx->aarch64_neon[1].v[0] +
+	    10 * ctx->aarch64_neon[1].v[1] + 20 * ctx->aarch64_neon[1].v[2] +
+	    34 * ctx->aarch64_neon[1].v[3] - 48 * ctx->aarch64_neon[2].v[0] -
+	    64 * ctx->aarch64_neon[2].v[1] - 80 * ctx->aarch64_neon[2].v[2] -
+	    96 * ctx->aarch64_neon[2].v[3] + 64 * ctx->aarch64_neon[3].v[0] +
+	    64 * ctx->aarch64_neon[3].v[1] + 64 * ctx->aarch64_neon[3].v[2] +
+	    64 * ctx->aarch64_neon[3].v[3];
+
 	ZIO_SET_CHECKSUM(zcp, A, B, C, D);
 }
 
-#define	NEON_INIT_LOOP()			\
-	asm("eor %[ZERO].16b,%[ZERO].16b,%[ZERO].16b\n"	\
-	"ld1 { %[ACC0].4s }, %[CTX0]\n"		\
-	"ld1 { %[ACC1].4s }, %[CTX1]\n"		\
-	"ld1 { %[ACC2].4s }, %[CTX2]\n"		\
-	"ld1 { %[ACC3].4s }, %[CTX3]\n"		\
-	: [ZERO] "=w" (ZERO),			\
-	[ACC0] "=w" (ACC0), [ACC1] "=w" (ACC1),	\
-	[ACC2] "=w" (ACC2), [ACC3] "=w" (ACC3)	\
-	: [CTX0] "Q" (ctx->aarch64_neon[0]),	\
-	[CTX1] "Q" (ctx->aarch64_neon[1]),	\
-	[CTX2] "Q" (ctx->aarch64_neon[2]),	\
+#define	PREF1X64L1(buffer, PREF_OFFSET, ITR) \
+	__asm__("PRFM PLDL1STRM, [%x[v],%[c]]"::[v]"r"(buffer), \
+	[c]"I"((PREF_OFFSET) + ((ITR) + 0)*64));
+
+#define	PREF1KL1(buffer, PREF_OFFSET) \
+	PREF1X64L1(buffer, (PREF_OFFSET), 0)
+
+#define	NEON_INIT_LOOP() \
+	asm("ld1 { %[AA2].4s, %[A3A4].4s }, %[CTX0]\n"          \
+	"ld1 { %[BB2].4s, %[B3B4].4s }, %[CTX1]\n"              \
+	"ld1 { %[CC2].4s, %[C3C4].4s }, %[CTX2]\n"              \
+	"ld1 { %[DD2].2d, %[D3D4].2d }, %[CTX3]\n"              \
```

Comment:
It is a pre-existing issue, but this likely is broken on big endian:

https://gist.github.com/lirenlin/8647388

`ldr` might be a better choice here. The two probably perform the same.

Reply:
@ryao Sorry for late response here. The ld1-ld4 are used for Neon register dedicated, and usually it's easy to use ld1-ld4 to operate 4 register once. But for ldr, easy command it can just operate one register[1]. So the ld1-ld4 can actually save some cpu cycle when load the data from memory.

In my opinion, we don't worry about the Big endian issue. The link you referred before is showing the issue for ld1 when it on big endian. Arm64 supports both little endian and big endian, but the endianess choice is according to the IMPLEMENT DEFINED[2], actually there is no real hardware for Arm64 big endian. The lack of user cases and  gaps for the upper layer software compliance prevent the Arm64BE. E.g, there is no Golang supported for Arm64BE[3].
1. https://developer.arm.com/documentation/102159/0400/Load-and-store---data-structures
2. https://developer.arm.com/documentation/102376/0100/Alignment-and-endianness
3. https://stackoverflow.com/questions/69174919/can-go-build-support-armbe-arm64be